File: Sieve_of_Eratosthenes/Go/jcla1/sieve.go
package main

import "math"

// Takes a limit and returns the
// number of primes below that limit
func Sieve(n int) int {
	// We use false for prime and
	// true for non-prime values in arr
	arr := make([]bool, n)
	for i := range arr {
		arr[i] = true
	}

	limit := int(math.Sqrt(float64(n)))

	for i := 2; i <= limit; i++ {
		if arr[i] {
			for j := i * i; j < n; j += i {
				arr[j] = false
			}
		}
	}

	counter := 0
	for _, p := range arr {
		if p {
			counter += 1
		}
	}

	// Account for numbers 0 & 1 being in the array
	return counter - 2
}

File: Sieve_of_Eratosthenes/Javascript/jcla1/sieve.js
// Returns the number of primes in the range [0, n)
function sieve(n) {
	var arr = new Array(n),
		limit = ~~Math.sqrt(n),
		counter = 0;

	// Isn't there an easier way to do this!?
	for (var i = arr.length - 1; i >= 0; i--) {
		arr[i] = true;
	}

	for (var i = 2; i <= limit; i++) {
		if (arr[i]) {
			for (var j = i*i; j < n; j += i) {
				arr[j] = false;
			}
		}
	}

	return arr.reduce(function(acc, b) {
		return b ? acc + 1 : acc;
	}, -2);
}
